File: routes/loans.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from web3 import Web3

router = APIRouter()

class LoanRequest(BaseModel):
    amount: float
    interestRate: float

# Configuración de la blockchain
provider = "https://mainnet.infura.io/v3/TU_INFURA_PROJECT_ID"
private_key = "TU_CLAVE_PRIVADA"
contract_address = "DIRECCIÓN_CONTRATO"
abi = [...]  # Copia el ABI del contrato

w3 = Web3(Web3.HTTPProvider(provider))
account = w3.eth.account.from_key(private_key)
contract = w3.eth.contract(address=contract_address, abi=abi)

@router.post("/api/loans/request")
async def request_loan(request: LoanRequest):
    try:
        tx = contract.functions.requestLoan(
            int(request.amount * 1e18),  # Convertir a Wei
            int(request.interestRate)
        ).buildTransaction({
            "from": account.address,
            "nonce": w3.eth.getTransactionCount(account.address),
            "gas": 2000000,
            "gasPrice": w3.toWei("20", "gwei")
        })
        signed_tx = w3.eth.account.sign_transaction(tx, private_key)
        tx_hash = w3.eth.sendRawTransaction(signed_tx.rawTransaction)
        return {"txHash": tx_hash.hex()}
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Error al solicitar préstamo: {str(e)}")

File: routes/loans.js
const express = require("express");
const { ethers } = require("ethers");
const router = express.Router();

const provider = new ethers.providers.JsonRpcProvider("https://YOUR_RPC_PROVIDER_URL");
const wallet = new ethers.Wallet("YOUR_PRIVATE_KEY", provider);
const contractAddress = "YOUR_CONTRACT_ADDRESS";
const abi = [
    // Inserta aquí el ABI de tu contrato inteligente
];
const contract = new ethers.Contract(contractAddress, abi, wallet);

router.post("/request-loan", async (req, res) => {
    const { amount, interestRate } = req.body;
    try {
        const tx = await contract.requestLoan(amount, interestRate);
        await tx.wait();
        res.status(200).send({ message: "Loan request submitted", txHash: tx.hash });
    } catch (error) {
        res.status(500).send({ error: error.message });
    }
});

module.exports = router;
